### PR TITLE
Implement per communicator universal identifiers for Firedrake objects

### LINKED ISSUE #3631: `utils._new_uid` should be collective over a communicator
As noted in #3628, the `_new_uid` facility can get out of sync across processes when multiple communicators are involved. Although most things in firedrake don't care about the uid, writing data _does_, so we should make sure that if the user doesn't provide names to functions, then `_new_uid` is safe even in a scenario with divergent processes as long as their program is correctly logically collective.

The way to handle this is to adapt `_new_uid` https://github.com/firedrakeproject/firedrake/blob/cb77d32ac8559d8d57fe0ef0efe251d9cbaf2309/firedrake/utils.py#L23

It should take a communicator (since it's only ever called in a context where a communicator is available), pull out the firedrake "internal" communicator (cc @JDBetteridge), and use that to key an id generation. This can be done by attaching an attribute to the communicator that is the `uid` counter. Since the calls to `_new_uid` are collective over a communicator, pulling the uid out of the communicator attribute and updating it will be safe in this scenario.

### PR DESCRIPTION
**Pull request by JDBetteridge**
# Description
This requires https://github.com/NGSolve/ngsPETSc/pull/41 to be merged simultaneously.

Fixes #3631 and potentially #3628.

### CODE REVIEW

File: firedrake/mesh.py [hunk 1]
```diff
@@ -2446,13 +2446,18 @@ def make_mesh_from_coordinates(coordinates, name, tolerance=0.5):
...
+        Communicator.
 
     Returns
     -------
     MeshGeometry
         The mesh.
 
     """
+    if comm is None:
+        raise ValueError("A comm must be provided when creating a mesh from coordinates")
```
Comment (wence-) [inline firedrake/mesh.py:2459]: Should you not be using the comm attached to the `coordinates`?

File: firedrake/utils.py [hunk 2]
```diff
@@ -6,9 +6,11 @@
 from pyop2.datatypes import RealType     # noqa: F401
 from pyop2.datatypes import IntType      # noqa: F401
 from pyop2.datatypes import as_ctypes    # noqa: F401
+from pyop2.mpi import COMM_WORLD
 from firedrake_configuration import get_config
 
-_current_uid = 0
+# MPI key value for storing a per communicator universal identifier
+FIREDRAKE_UID = COMM_WORLD.Create_keyval()
```
Comment (wence-) [inline firedrake/utils.py:13]:
```suggestion
from mpi4py import MPI
...
FIREDRAKE_UID = MPI.Comm.Create_keyval()
```

Then it's clear that the keyval is not specific to a comm?

**Review by connorjward — APPROVED** [review]
Looks fine to me.

**Comment by JDBetteridge** [comment]
> Didn't realise was failing CI.

It wasn't before, I need to resurrect my ngsPETSc branch, turns out I needed it after all...

**Review by dham — APPROVED** [review]

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -83,6 +83,7 @@ jobs:
             --install defcon \
             --install gadopt \
             --install asQ \
+            --package-branch ngsPETSc JDBetteridge/firedrake_uid_changes \
```
Comment (JDBetteridge) [inline .github/workflows/build.yml:86]:
```suggestion
```

**Review by connorjward — APPROVED** [review]

**Pull request merged**